### DISCUSSION TITLE
Question: Delete leftover XMP metadata when replacing images

### DISCUSSION THREAD

**Discussion by mnmtz** [py-pdf/pypdf#1928] [Q&A] [answered]
After successfully replacing images, the (old) references to XMP metadata object as well as the metadata objects still remain in the PDF file. How can I delete these objects?

Reference:
```
<< /BitsPerComponent 8 /ColorSpace 6 0 R /Filter /DCTDecode /Height 1759 /Intent /RelativeColorimetric /Metadata 150 0 R /Name /X /Subtype /Image /Type /XObject /Width 1245 /Length 813310 >>
```
XMP Object:
```
150 0 obj
<< /Subtype /XML /Type /Metadata /Length 2344 >>
stream
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="Adobe XMP Core 7.2-c000 79.1b65a79, 2022/06/13-17:46:14        ">
```

**Comment by pubpub-zz** [comment]
the easiest solution is just to delete the field entry: if obj is the object or an indirect object pointing the the object :
```python
obj = obj.get_object()
del obj["/Metadata"]
```

**Reply by mnmtz** [reply]
I have already tried that, but the xmp object still remains in the pdf file even if I delete the image. The only reference left is in the xref table...

**Reply by pubpub-zz** [reply]
if you want to clear the object try:
`writer._replace_object(obj.indirect_reference,NullObject())`

**Reply by mnmtz** [reply]
Thank you for the hint. Is there a way to get objects based on the position in the file structure or /Type?
According to the output from pdfalyzer, I would need to delete page[0]['/Resources']['/Metadata'], or am I wrong on this?

```
<160:Trailer(Dictionary)@/>
╠══ <3:Catalog(Dictionary)@/Root>
║   ╚══ <1:Pages(Dictionary)@/Root/Pages>
║       ╠══ <4:Page(Dictionary)@/Root/Pages/Kids[0]>
║       ║   ╠══ <5:Contents(EncodedStream)@/Root/Pages/Kids[0]/Contents>
║       ║   ╠══ <20:Thumb(EncodedStream)@/Root/Pages/Kids[0]/Thumb>
║       ║   ║   ╚══ <21:ColorSpace(Array)@/Root/Pages/Kids[0]/Thumb/ColorSpace>
║       ║   ║       ╚══ <22:(EncodedStream)@/Root/Pages/Kids[0]/Thumb/ColorSpace[3]>
║       ║   ╠══ <14:Resources(Dictionary)@/Root/Pages/Kids[0]/Resources[/Properties][/MC0]>
║       ║   ║   ╚══ <15:Metadata:XML(DecodedStream)@/Root/Pages/Kids[0]/Resources[/Properties][/MC0]/Metadata>
║       ║   ╠══ <16:Resources(Dictionary)@/Root/Pages/Kids[0]/Resources[/Properties][/MC1]>
║       ║   ║   ╚══ <17:Metadata:XML(DecodedStream)@/Root/Pages/Kids[0]/Resources[/Properties][/MC1]/Metadata>
║       ║   ╠══ <18:XObject:Image(EncodedStream)@/Root/Pages/Kids[0]/Resources[/XObject][/Im0]>
║       ║   ║   ╠══ <19:Metadata:XML(DecodedStream)@/Root/Pages/Kids[0]/Resources[/XObject][/Im0]/Metadata>
║       ║   ║   ╚══ /ColorSpace => <6:Resources[/ColorSpace][/CS0](Array)> (Non Child Reference)
║       ║   ╠══ /Resources[/ColorSpace][/CS0] => <6:Resources[/ColorSpace][/CS0](Array)> (Non Child Reference)
║       ║   ╠══ /Resources[/ExtGState][/GS0] => <8:ExtGState(Dictionary)> (Non Child Reference)
║       ║   ╚══ /Resources[/Font][/T1_0] => <9:Font:Type1(Dictionary)> (Non Child Reference)
```

**Reply by pubpub-zz** [reply]
objects can be saved as objects directly within an object (in a field of an objet of in an array) or stored within the catalog and then referenced through IndirectObjects allowing multiple reference to the same object. 
In the case you were describing before, deleting the entry was making the xmp StreamObject unreferenced and not usable by the PDF structure, although the data are still present : the viewer will then not use the xmp description 
I've provided you then with the solution to replace it with a null object. the entry being NullObject, the data are meaningless and ignored
finally Both solution can be combined if you prefer,making (150 0) object in the catalog a null object unreferenced. 
if you replace it with None instead of NullObject(), the object (150 0) will be completely ignored during catalog writing, but this may create some shift in the catalog and misalign the references and the catalog. I've not tried it and do not recommend this solution

**Reply by mnmtz** [reply]
Does it mean, that only the references, but not the object content (in this case sensitive information from the old image, that has been replaced) can be deleted and the content will always remain in the output file?

**Reply by pubpub-zz** [reply, marked as answer]
`writer._replace_object(obj.indirect_reference,NullObject())` will remove the sensitive information
